src/trace/harTracer.ts

```diff
@@ -0,0 +1,311 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as fs from 'fs';
+import * as util from 'util';
+import { BrowserContext, ContextListener, contextListeners } from '../server/browserContext';
+import { helper } from '../server/helper';
+import * as network from '../server/network';
+import { Page } from '../server/page';
+import * as har from './har';
+
+const fsWriteFileAsync = util.promisify(fs.writeFile.bind(fs));
+
+export function installHarTracer() {
+  contextListeners.add(new HarTracer());
+}
+
+class HarTracer implements ContextListener {
+  private _contextTracers = new Map<BrowserContext, HarContextTracer>();
```

Comment:
```suggestion
  private _contextTracers = new WeakMap<BrowserContext, HarContextTracer>();
```

Reply:
We instrument destruction and remove contexts from the map.

Reply:
yeah I got lost trying to make sure that when the browser is killed all of the contexts are destroyed